lib/semantics/check-do-stmt.cc

```diff
@@ -344,41 +372,108 @@ static CS GatherReferencesFromExpression(const parser::Expr &expression) {
   }
 }
 
-// Find a canonical DO CONCURRENT and enforce semantics checks on its body
-class DoConcurrentContext {
+// Find a DO statement and enforce semantics checks on its body
+class DoStmtContext {
 public:
-  DoConcurrentContext(SemanticsContext &context)
-    : messages_{context.messages()} {}
+  DoStmtContext(SemanticsContext &context)
+    : context_{context}, messages_{context.messages()} {}
 
-  bool operator==(const DoConcurrentContext &x) const { return this == &x; }
+  bool operator==(const DoStmtContext &x) const { return this == &x; }
 
   void Check(const parser::DoConstruct &doConstruct) {
+    if (doConstruct.IsDoConcurrent()) {
+      CheckDoConcurrent(doConstruct);
+      return;
+    }
+    if (doConstruct.IsDoNormal()) {
+      CheckDoNormal(doConstruct);
+      return;
+    }
+    // TODO: handle the other cases
+  }
+
+private:
+  using Bounds = parser::LoopControl::Bounds;
+
+  const Bounds &GetBounds(const parser::DoConstruct &doConstruct) {
     auto &doStmt{
         std::get<parser::Statement<parser::NonLabelDoStmt>>(doConstruct.t)};
     auto &optionalLoopControl{
         std::get<std::optional<parser::LoopControl>>(doStmt.statement.t)};
-    if (optionalLoopControl) {
-      currentStatementSourcePosition_ = doStmt.source;
-      if (auto *concurrent{std::get_if<parser::LoopControl::Concurrent>(
-              &optionalLoopControl->u)}) {
-        DoConcurrentEnforcement doConcurrentEnforcement{messages_};
-        parser::Walk(
-            std::get<parser::Block>(doConstruct.t), doConcurrentEnforcement);
-        DoConcurrentLabelEnforce doConcurrentLabelEnforce{messages_,
-            doConcurrentEnforcement.labels(), doConcurrentEnforcement.names(),
-            currentStatementSourcePosition_};
-        parser::Walk(
-            std::get<parser::Block>(doConstruct.t), doConcurrentLabelEnforce);
-        EnforceConcurrentLoopControl(*concurrent);
-      }
+    auto &loopControl{optionalLoopControl.value()};
+    return std::get<Bounds>(loopControl.u);
+  }
+
+  void CheckDoControl(parser::CharBlock sourceLocation, bool isReal) {
+    if ((isReal) && (!context_.warnOnNonstandardUsage())) {
+      // No messages for the default case
+    } else if (isReal && context_.warnOnNonstandardUsage() &&
+        (!context_.warningsAreErrors())) {
+      messages_.Say(sourceLocation,
+          "DO controls not of type INTEGER are a non-standard "
+          "extension"_en_US);
```

Comment:
The only other place we use `warnOnNonstandardUsage()` (in `expression.cc`) the message just says what is wrong, it doesn't say it is non-standard. We should consistently do it one way or the other.

Reply:
They're not standard *any more*.  They're an obsolete feature, not an extension.

Reply:
Sounds good.  Thanks for pointing this out.